Helper to bypass mock FS & expose real files/directories

nonara opened the pull request:
Adds two new features:

1) `mock.bypass()` - Allows executing commands against the actual FS
2) `mock.load()` - 
    Reads the filesystem to map actual files and directories from the filesystem

    Options: `recursive`, `lazyLoad`
    - If lazyLoad is true, file content isn't loaded until they're explicitly read. (keeps memory usage low, making it feasible to load large filesystems)

See readme for more detail

closes #62 
closes #213 
(possibly others?)

nonara commented:
Tests failed due to an issue Node 6 support (spread operator), and git didn't re-create an empty dir, which threw count off.

Fixed the code & improved the test in question.

nonara commented:
Will update README tomorrow and `@types/mock-fs` as soon as it the PR is approved. Would appreciate a comment to let me know it's been / being reviewed. 

Thanks!

nonara commented on lib/filesystem.js:
```diff
@@ -112,6 +112,8 @@ FileSystem.prototype.getItem = function(filepath) {
     if (item) {
       if (item instanceof Directory && name !== currentParts[i]) {
         // make sure traversal is allowed
+        // This fails for Windows directories which do not have execute permission, by default. It may be a good idea
```
The existing behaviour here might be of concern, later. Is there a better way to accommodate Windows?

I've added a workaround which can be seen [here](https://github.com/tschaub/mock-fs/pull/304/files#diff-6d186b954a58d5bb740f73d84fe39073R260).

nonara commented on lib/filesystem.js:
```diff
@@ -112,6 +112,8 @@ FileSystem.prototype.getItem = function(filepath) {
     if (item) {
       if (item instanceof Directory && name !== currentParts[i]) {
```
The existing behaviour below might be of concern, later. Is there a better way to accommodate Windows?

See notes on why and my workaround: [here](https://github.com/tschaub/mock-fs/pull/304/files#diff-6d186b954a58d5bb740f73d84fe39073R260).

nonara commented:
This can be considered complete (unless changes requested)

Recent updates:

- Made automatically created directories inherit privileges.
  - This lead to discovering an issue with windows permissions. (See my code comment above for detail)
- Updated readme

3cp commented on lib/index.js:
```diff
@@ -17,6 +17,15 @@ const realCreateWriteStream = fs.createWriteStream;
+
+const S_IXUSR = 0o100;
+const S_IXGRP = 0o10;
+const S_IXOTH = 0o1;
```
`lib/item.js` has these permissions defined but not exported, you can export them.

3cp commented on readme.md:
```diff
@@ -58,7 +60,51 @@ The second (optional) argument may include the properties below.
  * `createTmp` - `boolean` Create a directory for `os.tmpdir()`.  This is `true` by default.
 
-### Creating files
+### Automatically Creating Files & Directories
```
I suggest to clarify this API is to load real local dirs inside mock-fs's fake file system.
I have two doubts:
1. The API name is debatable.
2. I did not see the use cases of mixing mock such as:
```js
mockfs({ real_dir: ..., fake_dir: ... });

mockfs({fake_dir: {
  fake_sub_dir: {
    mount_a_real_dir: ...
    fake_file: ...
    another_fake_dir: ...
  }
});
```

I thought users were expecting the use cases like the above.

I guess your api support this?
```js
mockfs({ real_dir: mock.createDirectoryInfoFromPaths(...), fake_dir: ... });
```

nonara replied:
1. API Name

I followed convention which is generally adhered to for factories in TypeScript / JS code. You'll see this same format used in the TypeScript compiler, etc. The TypeScript type name of the config param that `mock()` takes is is `DirectoryInfo`, which is essentially an object that contains key: path, and value: content, DirectoryInfo, or a factory function which produces a `File` or `Directory` instance.

So this method creates such an object.

With that said, because this is a JS project, I agree that not all users may immediately recognize the name. Do you have a better name in mind?

2. Yes, you can easily merge automatic and manual

```
mock({
  ...mock.createDirectoryInfoFromPaths(...),
  'manual/path/': 'content'
});
```

I considered adding that to the readme. Should I?

3cp replied:
I might missunderstand your code. Yes, pls add that to readme.

If you simplify your api to read only one dir, then user can easily rename dir, and still easily load multiple dirs.

```js
mockfs({
  node_modules: mock.loadDir('../node_modules'),
  foo: mock.loadDir('bar')
  xx: {
    ...
  }
});
```

nonara replied:
I'll share my proposals followed by detail explaining them

1. Rename `mock.createDirectoryInfoFromPaths` to `mock.mapPaths()`
2. Add `mock.loadDir()`
3. Re-write the readme to clearly distinguish between automatic and manual mapping

## Automatic Mapping

The current factory generates a config object for full file paths. This is _automatic mapping_. The reason this is very useful is because it allows you to essentially create a shadow of specific parts of the actual file-system. Because it uses _lazy-loading_, we can now literally expose massive areas of your actual file-system without worrying that any modifications will take effect on the real hard drive.

This makes setting up the mock system as simple as `mock(mock.mapPaths([ path1, path2, path3 ]))`, and now all of those paths and their stat info is available.

## Manual mapping

For this, we have the existing functions, and I will also create a `mock.loadDir()`, which loads paths relatively and functions as you've shown.

What do you think of those proposals?

3cp replied:
I don't have strong opinion on the APIs, that sounds very good to me.
@tschaub may have better idea on what to do.

nonara replied:
Please dis-regard my previous comment. (I've deleted it)

## Current Proposal

1. Rename `mock.createDirectoryInfoFromPaths` to `mock.mapPaths()`
2. Add `mock.mapDir()` (same as your `mock.loadDir`)
3. Re-write the readme to clearly distinguish demonstrate automatic or manual mapping

## mock.mapPaths()

This is _automatic mapping_. It lets you create a full config map from one or more remote or relative paths.

Makes setting mock system as simple as `mock(mock.mapPaths('../'))` or `mock(mock.mapPaths([ '/srv/external-assets', '.' ])`, etc.

I'd recommend leaving this as it makes it simple and easy to quickly expose multiple 'mount points' on the filesystem. Due to lazy-loading, you can now expose very large areas.

## Manual mapping

This is for those who want to include some manual data. For this, we'll create `mock.mapDir()` that functions as you've shown.

```
mockfs({
  node_modules: mock.mapDir('../node_modules'),
  foo: mock.loadDir('bar', { recursive: false })
  xx: {
    ...
  }
});
```

What do you think on that?

nonara replied:
Great! I'll go ahead and make the changes for now. If @tschaub has any input, I'd be glad to hear it.

nonara commented:
Ok! I think you'll find the new API much more intuitive and flexible! I believe the new readme should be easy to understand, as well.

We now have three new 'Mapper Methods': `mock.mapPaths()`, `mock.mapDir()`, and `mock.mapFile()`.

I re-wrote everything which also allowed for better tests & better overall architecture.

3cp commented on lib/index.js:
```diff
@@ -183,3 +184,200 @@ exports.directory = FileSystem.directory;
+  }
+
+  // Deactivate mocked bindings
+  const binding = process.binding('fs')._mockedBinding;
```
`process.binding('fs')` can be replaced with `realBinding`.

3cp commented on lib/index.js:
```diff
@@ -183,3 +184,200 @@ exports.directory = FileSystem.directory;
+  delete process.binding('fs')._mockedBinding;
+
+  // Perform action
+  const res = fn();
```
need a try catch around this line to ensure mockedBinding is reactivated.

nonara commented:
Addressed comments.

tschaub commented on lib/index.js:
```diff
@@ -183,3 +184,203 @@ exports.directory = FileSystem.directory;
+  if (typeof p !== 'string') {
+    throw new TypeError(`Invalid path. All paths must be strings`);
+  }
+  return path.resolve(p);
```
`path.resolve(p)` already throws a `TypeError` if the argument is not a string.  I'm curious why this is needed.

nonara replied:
Simply to provide a specific error message that's makes the issue easily understood. Throwing custom TypeError with a more understandable message is a standard practice. I can remove if you'd prefer.

tschaub replied:
Yes, I'd prefer to have the error that people expect from `path.resolve()`.

tschaub commented on lib/index.js:
```diff
@@ -183,3 +184,203 @@ exports.directory = FileSystem.directory;
+ * Mapping
+ * ****************************************************************************************************************** */
+
+const mapping = {};
```
Instead of tacking properties onto this object, it would be more consistent with the rest of the repo to create a new module that `exports` the functions you want to expose.  Is there a reason these functions need to be defined in this same scope?

nonara replied:
I certainly prefer it that way! :) 

The items in question are helper functions meant to be used with functions that belong to the `mock` object. I attached them to the object to prevent polluting the module scope. 

I had planned on creating a new file, but because the exported functions are meant to be attached to `mock`, judging by the rest of the repo, I felt it may be breaking convention to create a new file for them. So this was done in effort to be courteous. I'm glad to move them to their own file.

tschaub commented on test/lib/index.spec.js:
```diff
@@ -180,6 +184,245 @@ describe('The API', function() {
+
+      assert.equal(fs.readFileSync('/path/to/file', 'utf8'), 'content');
+      assert.throws(() => fs.readFileSync(__filename));
+      assert.doesNotThrow(() => mock.bypass(() => fs.readFileSync(__filename)));
```
Instead of asserting that `mock.bypass` doesn't throw, I think it would be a more useful test to assert that the contents of a file read within a `mock.bypass` function are what we expect.  Or even just to assert that the return from `fs.access` (or `fs.exists`) is what we expect.

Maybe something like this
```js
let exists = false;
mock.bypass(() => {exists = fs.accessSync(__filename)});
assert.ok(exists);
```

nonara replied:
`accessSync` throws if it can't be accessed, also. Agreed on existsSync being more useful. I'll update it.

3cp commented:
There is a design issue. When user wants to bypass an async func, your implementation doesn't work.

```js
mock.bypass(async () => {
  const stats = await fs.promises.stat('/path/to/file.txt');
  ...
  const data = await fs.promises.readFile('path/to/file.txt');
  ...
});
```

Your implementation probably needs to wrap the return of `fn()` in `Promise.resolve(fn()` and use promise api to reset mock.

nonara commented:
> There is a design issue. When user wants to bypass an async func, your implementation doesn't work.

It was actually designed intentionally to be synchronous. Asynchronous code could introduce major problems if other code is attempting to access the file-system in parallel.

If you'd really like, I can make it support async and add a typescript JSDoc overload which includes a warning + add it to the readme.

3cp commented:
I also think async mode is troublesome, so I am not sure if my suggestion is a good idea. As long as you document it clearly.

nonara commented:
I will clarify in the readme that mock.bypass() should be used for synchronous calls. 

One possible route is to add `mock.enable()` and `mock.disable()` helpers for those who know what they're doing and want to risk async calls.

3cp commented:
Maybe keep `bypass()` func as internal? I think your other APIs can cover those github issues. @tschaub any thought?

Because, to end user, `bypass()` is not a show stopper. They can still do `mockfs.restore(); fn(); mockfs(fakeFs_again);` manually.

nonara commented:
> Because, to end user, `bypass()` is not a show stopper. They can still do `mockfs.restore(); fn(); mockfs(fakeFs_again);` manually.

Actually a lot of the reason I wanted to do the PR was because `bypass` is extremely helpful. If someone is pre-loading a large amount of data into the file-system, it's a hassle and can get expensive to re-create it all from scratch when I just need to do a quick in-line read from the real FS.

I highly recommend we keep it and just let people know that it's intended for synchronous calls.

nonara commented:
A good thing to keep in mind here is that because we now support lazy-loading, if you unmock, all your lazy load data gets wiped, which is certainly not ideal for tests situations where things need to be fast.

I think the suggestion of adding the advanced features `disable` and `enable` + clarifying documentation for `bypass` is clear and support a wide range of use cases. 

I'm not seeing a downside to adding the functionality, but I'm open to hearing it if you think that there are some?

nonara commented:
See latest for new readme which clarifies bypass methods and provides advanced methodology for users who know what they're doing.

nonara commented:
@tschaub Addressed your comments and made new modules.

tschaub commented on lib/bypass.js:
```diff
@@ -0,0 +1,53 @@
+
+  // Deactivate mocked bindings
+  exports.disable();
+  delete realBinding._mockedBinding;
```
It looks like this is redundant.

tschaub commented on lib/mapping.js:
```diff
@@ -0,0 +1,167 @@
+  }
+
+  stats = Object.assign({}, stats);
+  const outputPropKey = isRoot ? target : path.win32.basename(target);
```
Does this work on non-win32 systems?

nonara replied:
It does, yes. This ensures that basename works with _both_ posix and windows style paths. See [Windows vs. Posix](https://nodejs.org/api/path.html#path_windows_vs_posix).

Example:
```
> path.win32.basename('/a/b/c')
'c'
> path.win32.basename('C:\\windows\\c')
'c'
> path.posix.basename('/c/d')
'd'
> path.posix.basename('C:\\windows\\d')
'C:\\windows\\d'
```

For this use case, I'm not sure whether it would matter, though it might if the user was running from cygwin. Overall, it seemed best practice to use the most universally sound method.

3cp replied:
Plain `path.basename()` does the right thing for all paths when running on Windows too.

Your code has the advantage to support path `'C:\\windows\\c'` on posix systems. But if users use windows path name while testing code inside linux/mac, they will likely encounter issues in other places even if they have no issue here.

3cp replied:
I see your code is to load real dir, so it's irrelevant to support `'C:\\windows\\c'` on posix systems.
`path.basename` is enough.

nonara replied:
*sigh* 

This is simply best practice to standardize behaviour across all environments. As I stated, it may not be necessary, though in cygwin environments it could very likely cause issue. There is, however, zero downside to using it in posix, which is why it's best practice.

Posix result:
```
> path.basename('C:\\d\\e')
'C:\\d\\e'
> path.win32.basename('C:\\d\\e')
'e'
> path.basename('/c/d')
'd'
> path.win32.basename('/c/d')
'd'
>
```

3cp replied:
I see, I didn't test cygwin env.

3cp replied:
I really don't think it's a best practice due to there is difference in posix file name and windows file name.

In windows, `\` is not allowed in file name, but it's allowed in posix system.

In posix, `path.win32.basename` fields wrong result.

```⋊> ~/b/bcx-broker-au-frontend on master ◦ node                                                                09:00:24
Welcome to Node.js v12.18.2.
Type ".help" for more information.
> path.basename('/a/b\\c')
'b\\c'
> path.win32.basename('/a/b\\c')
'c'
>

3cp replied:
> To achieve consistent results when working with Windows file paths on any operating system, use path.win32:
To achieve consistent results when working with POSIX file paths on any operating system, use path.posix:

The doc didn't say use `path.win32` for consistent results with **both Windows and POSIX** file paths. I think what's the argument here.

nonara replied:
I'm aware of what it says. To be clear, when unsure of whether a path is windows or posix, it is the best practice with native libraries.

I was, however, unaware of the backslash issue. I'll remove the win32 and we'll hope cygwin doesn't take issue.

3cp replied:
cygwin always got weird issues. Better to test it out if you have the env.

On the other side, I always felt nodejs std lib should normalise the win32/posix path problem, many other languages do that.

nonara replied:
I agree! Thanks for pointing out the backslash issue.

tschaub commented on readme.md:
````diff
@@ -58,6 +60,70 @@ The second (optional) argument may include the properties below.
+  'fakefile.txt': 'content here'
+});
+```
+
````
I appreciate all that you've put together here, @nonara.

I have some concern about the size of the change to the API (adding `bypass`, `enable`, `disable`, `mapPaths`, `mapFile`, and `mapDir`).  And I think the subtle difference in usage of `mapPaths` vs `mapFile` or `mapDir` is error prone.

It feels like we could try to minimize the changes to the API while still providing most of the same functionality.

All of `mapPaths`, `mapFile`, and `mapDir` look to be about providing a convenient way to load up a mock filesystem with data from the real filesystem.  Words like `load`, `clone`, and `read` come to mind for function names.

```js
mock({
  '/path/to/dir': mock.load('/path/to/dir'),
  '/path/to/file': mock.load('/path/to/file')
})
```

That implies adding a single `load` function that takes either a path to a directory or a path to a file and copies contents and metadata into a corresponding mock filesystem entry.  I assume we should also make it work for symlinks.

If we find that usage patterns make it very awkward to repeat the real path in the mock path, we could add a function that makes that more convenient later.

The `bypass`, `enable`, and `disable` functions feel like they could be boil down to a single function that works with either an async/promise returning function or a sync function.  In the docs below, you warn against using async calls with the mock disabled - that same warning could be included in the `bypass` docs without adding the `enable` and `disable` functions.

Having a single `bypass` function also gets around the issue that calling `enable` twice makes `disable` not work (this is fixable, but could also be avoided by not adding these functions at all).

nonara replied:
> All of `mapPaths`, `mapFile`, and `mapDir` look to be about providing a convenient way to load up a mock filesystem with data from the real filesystem. Words like `load`, `clone`, and `read` come to mind for function names.

I agree that `mapDir` and `mapFile` could be merged, and `load` is a great name.

As for mapPaths, this is a simple way to quickly load multiple real paths.

```ts
// With (simple)
mock(mock.loadPaths(realPaths));

// Without (simple)
mock(realPaths.reduce((p,c) => { p[c] = map.load(p) }, {})));

// With (extended)
mock{{
  ...mock.loadPaths(realPaths),
  'extra': 'content'
});

// Without (extended)
mock({
  ...realPaths.reduce((p,c) => { p[c] = map.load(p) }, {})),
  'extra': 'content'
});
```   

Bear in mind that while reduce is ugly and difficult to read, this is the shortest solution. Most people will end up using multi-line for loops, etc, all to accomplish what many have requested - a feature to simply 'mount' real areas. Worse yet, some will end up doing

```
{ 
  '/path/to/my/location': map.load('/path/to/my/location'),
  '/path/to/other.location': map.load('/path/to/other/location'),
  ...
}
```

Tedious.

The decision lies with you. I really have spent more time on this than I am able, so I'll go with whatever you want, but I hope you'll consider it reasonable to allow the API to have three new items: `load`, `bypass`, and `loadPaths`

Otherwise, many (myself included) will end up replicating helper logic to convert an array of real locations for each package we use this in.

> The `bypass`, `enable`, and `disable` functions feel like they could be boiled down to a single function that works with either an async/promise returning function or a sync function. In the docs below, you warn against using async calls with the mock disabled - that same warning could be included in the `bypass` docs without adding the `enable` and `disable` functions.

Agreed. The consideration was that many people will simply use the function without noticing the warning. It was set up to be more strict to prevent people using it and filing issues. 

But this is entirely your call, if you're good with it, then it's no problem. I'll add async support, and we can add the warning to both the readme and JSDoc for the types package to make sure it's seen.

> Having a single `bypass` function also gets around the issue that calling `enable` twice makes `disable` not work (this is fixable, but could also be avoided by not adding these functions at all).

Didn't see that. I was up quite late last night. I'll correct the issue. If you're alright with it, I'd like to leave them attached to exports and simply not document in readme or in the @types package. That way there is zero-footprint, but people who know the source can use them. 

I can always replicate the behaviour to do it, but this way if anything changes my code doesn't break.

Let me know! Hopefully we can get this wrapped up. I'll wait until your responses before updating anything.

3cp replied:
I don't feel tediousness is an issue. I cannot imagine people use mock-fs to load lots of real dirs, it seems defeating the purpose of mocking.

nonara replied:
> I cannot imagine people use mock-fs to load lots of real dirs, it seems defeating the purpose of mocking.

Not at all. If we have multiple assets directories or even hard-coded paths on the system, this allows `mock-fs` to be used in a way that lets us 'mount' those areas and do whatever we like without actually modifying them. This is tremendously useful with a broad range of use-cases. 

That is, in fact, the reason I started this PR, but that's really all I can say on that. I don't want to belabor it further.

3cp replied:
> without actually modifying them

That's a good use case.

nonara commented:
### Updated according to last review:

- Remove `enable`, `disable` from exports
- Add promise support to `mock.bypass()`
- Merge `mapDir()` and `mapFile()` functions to `load()`
- Remove `mapPaths()`
- Use `path.basename` (May affect cygwin, but Linux allows backslashes in filename, which takes precedence)
- Remove custom error message for non-string parameter to `load()`
- Updated tests to accommodate changes
- Updated readme according to suggestions

### Note

`mock.bypass()` is built to work with Promises, however, async callback functions won't work. That was part of the idea behind `enable`, `disable`.

Is this acceptable? Would you like to alter the readme or handle it somehow?

nonara commented:
Investigating the TypeError with Buffer in Windows Node 12, now. At a bit of a loss on this. I can't see how any changes would have affected that. Will debug offline.

3cp commented:
> `mock.bypass()` is built to work with Promises, however, async callback functions won't work. That was part of the idea behind `enable`, `disable`.

I am quite sure your new implementation supports `mock.bypass(async () => {...});`, which you can add to unit tests.
Async return value is a promise.

3cp commented on lib/bypass.js:
```diff
@@ -0,0 +1,55 @@
+    const res = fn();
+
+    // Handle promise return
+    if (res.then) {
```
Suggest to check typeof function.

nonara commented:
> I am quite sure your new implementation supports `await mock.bypass(async () => {...});`, which you can add to unit tests.
> Async return value is a promise.

Should clarify - I'm referring to `fs` functions which utilize callbacks, like `fs.stat()`. Any further fs access within the callback would not work.

nonara commented:
I just checked out the **current source** from upstream master and confirmed that the buffer TypeError is happening on Windows for Node 12.

`TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string. Received an instance of Buffer`

3cp commented:
For the buffer thing, can you find the exact line?

Nodejs fs APIs support string/buffer on path, for example in the doc:
```
fs.access(path[, mode], callback)[src]#

History
path <string> | <Buffer> | <URL>
```

But Nodejs path APIs only support string.
```
path.basename(path[, ext])#

History
path <string>
```

This means if you forward a `path` got from a fs API to a path API, you need to take care of the buffer thing. I believe you can find there are some treatments in the existing code.

nonara commented:
It's in regard to `fs.symlink()` and is happening on master. I don't have a Windows Node 12 environment setup, I just created a new branch to test GH Actions to verify.

Error:

```
  1) fs.symlink(srcpath, dstpath, [type], callback)
       supports Buffer input:
     TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string. Received an instance of Buffer
      at validateString (internal/validators.js:120:11)
      at Object.isAbsolute (path.js:353:5)
      at preprocessSymlinkDestination (internal/fs/utils.js:317:18)
      at Object.symlink (fs.js:1073:19)
      at Context.<anonymous> (test\lib\fs.link-symlink.spec.js:192:8)
      at processImmediate (internal/timers.js:456:21)
```

Code affected:

https://github.com/tschaub/mock-fs/blob/master/test/lib/fs.link-symlink.spec.js#L191-L204

3cp commented:
I see, could be a regression related to latest nodejs v12.18.3.

3cp commented:
Confirmed it's a win32 only regression in latest nodejs v12, I will try to investigate.

I think we can ignore that failure in this PR.

nonara commented:
@3cp How'd it go? Any luck on determining a solution?

3cp commented:
I have not spent time on the win32 problem yet. Will do in coming week.

tschaub commented on lib/bypass.js:
```diff
@@ -0,0 +1,55 @@
+    // Handle promise return
+    if (typeof res.then === 'function') {
+      res.then(exports.enable);
+      res.catch(exports.enable);
```
This would be better handled with `res.finally(exports.enable)`.

tschaub commented on lib/bypass.js:
```diff
@@ -0,0 +1,55 @@
+    return res;
+  } catch (e) {
+    exports.enable();
+    throw e;
```
This could be in a `finally` block.

tschaub commented on readme.md:
```diff
@@ -58,6 +60,42 @@ The second (optional) argument may include the properties below.
+
+| Option    | Type    | Default | Description |
+| --------- | ------- | ------- | ------------
+| lazyLoad  | boolean | true    | File content isn't loaded until explicitly read
```
I think this could just be called `lazy` - avoiding a bit of repetition: `load(dir, {lazy: true})`.

nonara commented:
> Thanks for this great contribution @nonara.

Sure thing! Thanks for taking the time to review.

3cp commented:
The win32 issue is indeed a regression in nodejs code base. I will create a PR or issue there.

3cp commented:
The win32 issue is fixed in nodejs, but not patched in v12 yet. https://github.com/nodejs/node/issues/34514

Johnhhorton commented:
Any idea when this will be released? Working on a project that could really use it ❤️

3cp commented:
There are some edge cases to be cleaned up (https://github.com/tschaub/mock-fs/pull/306#issuecomment-671129346). I will PR the fix shortly after merging #303.

dqisme commented:
Hi, thanks for this wonderful upgrade
when will we have this feature in the next release, please?

frederikheld commented:
@nonara Many thanks for your effort! That's a great addition to mockFs that will make testing much easier :-)

nonara commented:
@frederikheld You're very welcome. Thank you for the kind comment!
